core/src/main/java/com/google/cloud/sql/core/RateLimitCalculator.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.sql.core;
+
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A simple, constant-time rate limit calculator. Ensures that there is always at least
+ * delayBetweenAttempts milliseconds between attempts.
+ */
+class RateLimitCalculator {
+  private volatile long nextOperationTimestamp;
+  private final long delayBetweenAttempts;
+
+  RateLimitCalculator(long delayBetweenAttempts) {
+    this.delayBetweenAttempts = delayBetweenAttempts;
+  }
+
+  /**
+   * Returns the number of milliseconds to delay before proceeding with the rate limited operation.
+   * If this returns <= 0, the operation must call "acquire" again until it returns -1 to avoid race
+   * conditions.
+   */
+  synchronized long acquire() {
```

Comment:
Shouldn't this be a private method?

Reply:
Can we test only through the public interface? Would a deterministic scheduler help do this? We should try to avoid testing private interfaces.

Reply:
Unfortunately, this is one of those rare situations where we need to unit test a private method because: 

- there are detailed states that we must test precisely to ensure that the rate limiter works.
- the rate limiter's public API does not grant the precision we need to create those states reliably. 
 
I can't write a precise test case using the public API `acquireAsync()` to validate that the delay calculation logic is working properly. Timing in a real thread pool is too unreliable. The jmock DeterministicScheduler doesn't work correctly with scheduled tasks in a ListeningScheduledExecutorService.